File: CPP/ByTier/F/1869_F_Flower_like_Pseudotree.cpp
// Problem: CF 1869 F - Flower-like Pseudotree
// https://codeforces.com/contest/1869/problem/F

/* 
 * Flower-like Pseudotree Construction
 * 
 * Problem: Given vertex degrees, construct a flower-like pseudotree.
 * 
 * A flower-like pseudotree is a connected graph with exactly one cycle and no self-loops, 
 * such that after removing the cycle edges, the resulting forest has all trees with 
 * equal depth when rooted at their corresponding cycle vertex.
 * 
 * Key Observations:
 * - A pseudotree with n vertices has exactly n edges.
 * - The total degree sum must be 2 * n for a valid pseudotree.
 * - For a flower-like pseudotree, certain conditions involving vertex degrees determine 
 *   if construction is possible.
 * 
 * Algorithm:
 * 1. Preprocess degrees and sort them.
 * 2. Classify vertices into degrees 1, 2, and others.
 * 3. Use different strategies to build the structure based on these classifications.
 * 4. For valid constructions, connect leaves and build cycles appropriately to ensure
 *    flower-like property.
 * 
 * Time Complexity: O(n log n) - due to sorting and linear processing
 * Space Complexity: O(n) - for storing degrees and edges
 */
#include <stdio.h>
#include <algorithm>
#include <cstring>
#include <iostream>
#include <utility>

using namespace std;
const int N = 1e6 + 10;
int n;
pair<int, int> d[N];
void connect(int a, int b) {
  // Output an edge between vertices a and b (based on their indices)
  printf("%d %d\n", d[a].second, d[b].second);
  d[a].first--;
  d[b].first--;
}
void createCycle(int start, int end) {
  // Create a cycle from start to end vertices (indices)
  for (int i = start; i < end; i++) {
    connect(i, i + 1);
  }
  connect(end, start);
}
void solveLeaves(int totalLeaves) {
  // Connect leaves to the cycle, ensuring each has appropriate degree
  for (int i = 0, j = totalLeaves; i < totalLeaves; i++) {
    while (d[j].first == 0)
      j++;
    connect(i, j);
  }
}
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int cases;
  scanf("%d", &cases);
  while (cases--) {
    scanf("%d", &n);
    long long sum = 0;
    for (int i = 0; i < n; i++) {
      scanf("%d", &d[i].first);
      d[i].second = i + 1;
      sum += d[i].first;
    }
    sort(d, d + n); // Sort vertices by degree to group them efficiently
    
    int ones = 0, twos = 0;
    while (ones < n && d[ones].first == 1)
      ones++;
    while (ones + twos < n && d[ones + twos].first == 2)
      twos++;
    int left = n - ones - twos; // Count of vertices with degree >= 3
    
    if (sum != 2 * n || left == 1) {
      // Check for basic validitiy:
      // 1. Total degree sum must equal 2 * n
      // 2. If there's only one vertex with degree >= 3, cannot form valid flower-like pseudotree
      printf("No\n");
    } else if (ones == 0 || twos == 0) {
      // If there are no vertices of degree 1 or 2, construct simple cycle with leaves
      printf("Yes\n");
      createCycle(ones, n - 1);
      solveLeaves(ones);
    } else if ((left + twos) % 2 == 0) {
      // If the combined count of higher-degree vertices and 2s is even
      printf("Yes\n");
      createCycle(n - 2, n - 1);
      int nxt = n - 1, nxtnxt = n - 2;
      for (int i = n - 3; i >= ones; i--) {
        connect(i, nxt);
        nxt = nxtnxt;
        nxtnxt = i;
      }
      solveLeaves(ones);
    } else if (twos + left < 5) {
      // Not enough vertices to form a valid structure
      printf("No\n");
    } else if (d[n - 1].first >= 4) {
      // If the highest degree is at least 4
      printf("Yes\n");
      createCycle(n - 2, n - 1);
      int nxt = n - 1, nxtnxt = n - 2;
      connect(n - 3, n - 1);
      for (int i = n - 4; i >= ones; i--) {
        connect(i, nxt);
        nxt = nxtnxt;
        nxtnxt = i;
      }
      solveLeaves(ones);
    } else if (left >= 3 && left + twos >= 7) {
      // If at least 3 vertices have degree >= 3 and total >= 7
      printf("Yes\n");
      createCycle(n - 2, n - 1);
      connect(n - 3, n - 1);
      connect(n - 4, n - 2);
      connect(n - 5, n - 3);
      int nxt = n - 3, nxtnxt = n - 4;
      for (int i = n - 6; i >= ones; i--) {
        connect(i, nxt);
        nxt = nxtnxt;
        nxtnxt = i;
      }
      solveLeaves(ones);
    } else {
      // If no condition applies, construction is impossible
      printf("No\n");
    }
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1869/1869_F_Flower_like_Pseudotree.py
# Problem: CF 1869 F - Flower-like Pseudotree
# https://codeforces.com/contest/1869/problem/F

"""
Code Purpose:
This code determines whether a flower-like pseudotree can be constructed given
the degrees of vertices. A flower-like pseudotree is a connected graph with
exactly one cycle and all trees in the forest formed by removing the cycle
have the same depth when rooted at the vertex on the cycle.

Algorithms/Techniques:
- Graph theory and degree sequence analysis
- Construction algorithm based on sorting vertices by degree
- Greedy edge pairing to form a valid pseudotree structure

Time Complexity:
O(n log n) per test case due to sorting operations, where n is the number of vertices.
Space Complexity:
O(n) for storing the degree array and result edges.
"""

import sys


def find_pairs(ds):
    n = len(ds)
    if sum(ds) != 2 * n:
        return None
    if 2 not in ds:
        # If no vertex has degree 2, process as a special case
        hi, lo = [], []
        for i, d in enumerate(ds):
            if d > 2:
                hi += [i]
            else:
                lo += [i]
        ans = list(zip(hi, hi[1:] + hi[:1]))
        for i in hi:
            for j in range(ds[i] - 2):
                ans += [(i, lo.pop())]
        return ans
    big = sum(v > 2 for v in ds)
    if big == 0:
        # All vertices have degree <= 2, form a cycle
        a = list(range(n))
        return list(zip(a, a[1:] + a[:1]))
    if big == 1 or (big == 2 and ds.count(2) == 1):
        # Special cases where construction is impossible
        return None
    if max(ds) == 3:
        a, b = ds.count(3), ds.count(2)
        if (a + b) % 2 == 1 and (a == 2 or a + b < 6):
            return None
    ans = []

    def add_edge(i, j):
        ans.append((i, j))
        ds[i] -= 1
        ds[j] -= 1

    # Sort vertices by degree in descending order
    ids = sorted(range(n), key=lambda i: ds[i], reverse=True)
    add_edge(ids[0], ids[1])
    add_edge(ids[1], ids[0])
    for i in range(2, n - 1, 2):
        if ds[ids[i + 1]] < 2:
            break
        add_edge(ids[i - 2], ids[i])
        add_edge(ids[i - 1], ids[i + 1])
    hot = ids[:i][::-1]
    for u in ids[i:]:
        while ds[hot[-1]] == 0:
            hot.pop()
        add_edge(u, hot[-1])
        hot += [u]
    return ans


input_func = lambda: sys.stdin.readline()
output_func = lambda *args: sys.stdout.write(" ".join(map(str, args)) + "\n")
for _ in range(int(input_func())):
    n = int(input_func())
    ans = find_pairs(list(map(int, input_func().split())))
    if not ans:
        output_func("No")
        continue
    output_func("Yes")
    for i, j in ans:
        output_func(i + 1, j + 1)


# https://github.com/VaHiX/CodeForces/